File: 19. Remove Nth Node From End of List/removeNthFromEnd.js
// Definition for singly-linked list.
class ListNode {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
const removeNthFromEnd = (head, n) => {
  const dummyHead = new ListNode(0, head);
  let slow = dummyHead;
  let fast = head;

  while (n > 0 && fast) {
    fast = fast.next;
    n--;
  }

  while (fast) {
    slow = slow.next;
    fast = fast.next;
  }

  slow.next = slow.next.next;

  return dummyHead.next;
};


const head = new ListNode(1);
head.next = new ListNode(2);
head.next.next = new ListNode(3);
head.next.next.next = new ListNode(4);
head.next.next.next.next = new ListNode(5);

console.log('removeNthFromEnd: ', removeNthFromEnd(head, 2));

File: 19. Remove Nth Node From End of List/removeNthFromEnd.py
# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next

class Solution:
  def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
    dummyNode = ListNode(0, head)
    fast, slow = head, dummyNode

    while n > 0 and fast:
      fast = fast.next
      n -= 1

    while fast:
      fast = fast.next
      slow = slow.next

    slow.next = slow.next.next

    return dummyNode.next


head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
head.next.next.next.next = ListNode(5)

solution = Solution()
print('removeNthFromEnd: ', solution.removeNthFromEnd(head, 2))
